Backport diagnostic reporting to Lean 3.

Julian opened the pull request:
Does so by introducing infoview components (which in the future
possibly is a way to customize what's in the infoview display).

Closes: #41
Refs: #17

@gebner if you happen to have a spare moment I wouldn't mind a sanity check on whether [this crude splitting of Lean 3 goal state](https://github.com/Julian/lean.nvim/pull/70/files#diff-11f5cd6ffbb1e3a2d6e4a3948393a130292fd3d0f7e075249a20d6ebcafdb70cR19) is trivially incorrect in some way. It seems to work for the files / state I checked manually. If not no worries, can always revert later if it goes sideways.

rish987 commented:
Thanks for cleaning up `infoview.lua`. This is interesting, I'm curious do you think it would be possible to do this by adding a new type of provider to [to `lean-language-server`](https://github.com/leanprover/lean-client-js/blob/520e4bb4b202c988ea496bb3994f9be164c7f853/lean-language-server/src/index.ts#L156) that does the parsing rather than doing that here?

gebner commented on lua/lean/lean3.lua:
```diff
@@ -1,22 +1,52 @@
 
+-- Split a Lean 3 server response on goals.
+--
+-- Looks for ⊢, but ignores indented following lines for multi-line
```
That's the correct heuristic.  (⊢ at the beginning of a line until an empty line would work as well)

gebner commented on lua/tests/infoview/lsp_spec.lua:
```diff
@@ -34,7 +34,7 @@ describe('infoview', function()
     function(_)
       local text = infoview_lsp_update({3, 23})
-      assert.has_all(text, {"⊢ ℕ"})
+      assert.has_all(text, {"▶ goals accomplished 🎉"})
```
The result now means something completely different (previously 1 goal, now 0 goals).

Julian replied:
Ech, yes, will have to check what's happening here, I changed it too mindlessly.

Julian replied:
Fixed, this was me not properly checking that the goal response had any goals in it...

gebner commented on syntax/leaninfo.vim:
```diff
@@ -5,9 +5,9 @@ syn match leanInfoGoalVDash "^⊢"
-syn match leanInfoError "^▶.*: error:$"
-syn match leanInfoWarning "^▶.*: warning:$"
-syn match leanInfoInfo "^▶.*: information:$"
+syn match leanInfoError "^▶.*: Error:$"
```
I have to admit I personally prefer lowercase titles (because they don't stand out that much visually), but I'm willing to compromise on that.

However, the titles should stay consistent.  If "Error" is now capitalized, then "Expected type"  (and "Goals accomplished") should be capitalized as well.

Julian replied:
Fair, agreed, reverted!

gebner commented on lua/lean/infoview/components.lua:
```diff
@@ -0,0 +1,84 @@
+local components = {}
+
+local _SEVERITY = vim.tbl_extend("error",
+  { [0] = "Other" },
```
If you want to reuse the built-in DiagnosticSeverity map, then you can drop the Other case.  I've only included it for completeness, but it's not used by Lean afaict.

Julian replied:
Done.

Julian commented on lua/lean/infoview/components.lua:
```diff
@@ -0,0 +1,84 @@
+---plainGoal request
+---@param term_goal table: a term goal LSP response as returned by the Lean *4*
+---plainTermGoal request
+function components.combined_goals(goal, term_goal)
```
Yep, reasonable. Done.

Julian commented on doc/lean.txt:
```diff
@@ -43,6 +43,36 @@ infoview.get_info_lines()                          *infoview.get_info_lines()*
+
+
+    Parameters: ~
+        {goal} (table)  a goal LSP response as returned by the Lean *4*
```
Yeah I was fighting with fixing that before my call... I can't figure it out at the minute, even though I see it's expected to work from tjdevries/tree-sitter-lua#14, so I pinged the maintainers.

For now I just shortened the docstring...

Julian commented:
> Thanks for cleaning up infoview.lua. This is interesting, I'm curious do you think it would be possible to do this by adding a new type of provider to lean-language-server that does the parsing rather than doing that here?
I think it'd be possible but it still wouldn't be 100% seamless, there's still some heuristic'ing because I think the upstream Lean 3 server message doesn't separate things out. And combined with knowing the lean-language-server repo isn't super active I thought it was easier to just shove here, though now Gabriel's merged the outstanding PR I had there.

Julian commented:
Dah, well done @rish987 -- I added `is_empty` precisely because I thought it'd fix it (and changed where you changed), but forgot the `table.concat`... *le facepalm*.

Appreciated!

rish987 commented:
Fixed `infoview/lsp_spec.lua` LSP wait condition. LGTM at this point!
